### ISSUE TITLE
Erroneous local variable type warnings

### ISSUE THREAD

**Issue by jaylawl** [SkriptLang/Skript#1379]
... well, i'm beat to be honest:
https://snag.gy/I0pVx7.jpg
https://snag.gy/3sgMBl.jpg

I can't scroll back far enough to read all these messages. Everything was working just fine before upgrading, no errors. Couldn't find this change in the changelog - what's going on?

**Comment by Nicofisi** [comment]
Send the full script or anything that can provide more context - where are these variables set/changed/used?

**Comment by jaylawl** [comment]
```
        if hasTagValue(victim, "npc-ball") = true:
            set {_owner} to getTagValue(victim, "npc-ball")
            loop all entities in world "golf":
                loop-entity's uuid = {_owner}
                set {_owner} to loop-entity
                exit loop
            distance between {_owner} and victim < 3
```
It's complaining about this for example `variable is entity, not location`

**Comment by TheBentoBox** [comment]
Setting this to high prio since it seems pretty problematic.

@bensku I believe the issues people are experiencing are due to the new local variable type check errors not supporting default converters. Not sure if that's really fixable (especially when it comes to converters from addons) but if not then I think the functionality needs to be removed since it's giving errors where it simply shouldn't (like reported here, blocks should be perfectly acceptable here).

**Comment by Nicofisi** [comment]
I'd say that it's something pretty critical and it'd be perfect if a new release with only a fix was released within the next few days

**Comment by Pikachu920** [comment]
i really don't think variable type checking is a good idea

**Comment by Nicofisi** [comment]
Actually it's only a warning, I don't think this prevents scripts from loading - https://github.com/SkriptLang/Skript/blob/4edc5b09fae20b7aea2147a5432b27a00aea7f59/src/main/java/ch/njol/skript/lang/Variable.java#L176-L178

If it really doesn't then the issue probably isn't *that* critical, but in reality I think it still is

**Comment by jaylawl** [comment]
It did break things for me.

**Comment by TheLimeGlass** [comment]
The Type hints don't check converters/changers so everything that does will error.

**Comment by Pikachu920** [comment]
imo, type hints should be disabled for now until they are smarter (e.g. recognizing a variable as having multiple potential types and respecting converters and object expressions being assumed as correct). Even then it should still be an option really as it's becoming increasingly common for addons to manipulate variables in ways skript can't predict. i wouldn't be opposed to removing them completely, either. skript really isn't a typed language.

**Comment by fl-y** [comment]
https://pastebin.com/kT7Ry9i7
Used to have 0 errors
After update, 356 warnings, and most stuff broken.
Currently unusable.

**Comment by bensku** [comment]
Note that type hints produce *warnings*, not errors. If scripts are failing, it is likely that something else is broken.

**Comment by jaylawl** [comment]
Makes sense. I was also having errors, but i have no way of reading them because they're spammed away by the warnings. Unfortunately my server's console only shows a limited portion of the latest log.

**Comment by bensku** [comment]
Need a test case without addons. I wrote a simple one myself, but it does not trigger the bug in release build, so I have no idea if my fix is working.

**Comment by jaylawl** [comment]
Found some more (marked with ####):
https://snag.gy/nzPYcv.jpg
```
every second:

    loop all zombies in world "golf":
        delete {_npc}, {_ball}, {_hole}, {_vector}, {_destination} and {_potential-balls::*}
        hasTag(loop-entity, "jnpc-behaviour:putting") = true
        set {_npc} to loop-entity
        
        loop all chickens in world "golf":
            hasTag(loop-entity-2, "ball-type:putting-practice") = true
            if hasTag(loop-entity-2, "npc-ball:%{_npc}'s uuid%") = true:
                set {_ball} to loop-entity-2
                exit loop
            else:
                distance between {_npc} and loop-entity-2 < 16 ######################
                hasTagValue(loop-entity-2, "npc-ball") = false
                add loop-entity-2 to {_potential-balls::*}
                
        if {_ball} is not set:
            {_potential-balls::*} is set
            set {_ball} to random element out of {_potential-balls::*}
            addTag({_ball}, "npc-ball:%{_npc}'s uuid%")
            jnpc_findPuttingHole({_npc}, {_ball})
            
        if {_ball} is set:
            (x of {_ball}'s velocity + z of {_ball}'s velocity) = 0
            if getTagValue({_npc}, "npc-hole-x") is not set:
                jnpc_findPuttingHole({_npc}, {_ball})
            getTagValue({_npc}, "npc-hole-x") is set
            set {_x} to getTagValue({_npc}, "npc-hole-x")
            set {_z} to getTagValue({_npc}, "npc-hole-z")
            set {_hole} to location(("%{_x}%" parsed as number), y-coord of {_ball}'s location, ("%{_z}%" parsed as number), {_ball}'s world)
        
        set {_last-action} to getTagValue({_npc}, "last-action")
        if {_last-action} is not set:
            set {_last-action} to "unknown"
        if {_ball} and {_hole} are set:
            if {_last-action} = "prepare_stroke":
                setTagValue({_npc}, "last-action", "stroke")
                distance between {_npc} and {_ball} < 3 ######################
                (x of {_npc}'s velocity + z of {_npc}'s velocity) = 0
                set {_npc}'s target to {_ball}
         [...]
```

**Comment by TheBentoBox** [comment]
Looks like it's doing type checking even for unset variables, which it shouldn't do.

**Comment by jaylawl** [comment]
Apart from those warnings however, everything seems to be working fine. The only notable thing that's broken now is TusKe, which is a shame because my gut feeling says TukeNuke won't work on it anymore. But that's another story. tl;dr: seems stable

**Comment by TheBentoBox** [comment]
That's great to hear, thanks for the report. We'll be working through those false warnings (and if it seems to hard to resolve them all, bensku will just remove that new feature).

I will message Tuke personally but I'm not sure that will help anything. Can't hurt though.

**Comment by TheBentoBox** [comment]
The scripts *are* meant to work correctly regardless; that's why they're warnings, not errors. But thank you for the extra info. That one with the world variable is weird in particular since string -> world is a converter which should be covered by the recent fix.

**Comment by fl-y** [comment]
Tried out the bugfix version.
Doesn't spam the console at least... but a lot of things that used to work with the previous version simply don't work.
I understand that it's common courtesy to specify exactly which part of my skripts are not working, but I currently use more than 500kb of skript and thus simply don't have the time to find and pinpoint exactly where.
The weird thing is that skript isn't throwing a single error but simply doesn't work with most mechanics that I created. Warnings still spam when I reload skript though(pretty much the same amount as before)

**Comment by TheBentoBox** [comment]
> but I currently use more than 500kb of skript and thus simply don't have the time to find and pinpoint exactly where.

At least give the error messages, my guy. If you're going to provided quite literally nothing then the comment is useless. Not trying to be rude there, but we really can't do **anything** with that amount of info (i.e. zero, "there are errors" is not valuable info).

If you want things fixed, you can't just say you "don't have time" to properly report errors - you'll need to find the time :)

**Comment by TheBentoBox** [comment]
@jaylawl and @SlimeDog how's dev37c perform for you? Bensku released it this morning with even more type hint fixes. It should address most everything now.

**Comment by SlimeDog** [comment]
Just returned home, loaded up dev37c, and tested. Looks good. All the warnings on my scripts are gone. Thanks.

**Comment by jaylawl** [comment]
Looks good here aswell!

**Comment by bensku** [comment]
Finally!

**Comment by jaylawl** [comment]
More of these errors: https://snag.gy/UXiHoW.jpg
Code: https://hastebin.com/foxetiwigi.makefile

**Comment by Pikachu920** [comment]
are you sure this isn't related to #1394?

**Comment by jaylawl** [comment]
to be honest i don't know which it is...

**Comment by Pikachu920** [comment]
is it in a function? I'll reopen for now until @bensku responds

**Comment by jaylawl** [comment]
Yeah, both of which i posted are within functions

**Comment by Pikachu920** [comment]
ok, it's likely related to #1394 then but i will still let bensku take a look at this before it's closed for good
